src/cockpit/router.py

```diff
@@ -61,6 +61,7 @@ class Endpoint:
     __endpoint_frozen_queue: Optional[ExecutionQueue] = None
 
     def __init__(self, router: 'Router'):
+        router.endpoints[self] = set()
```

Comment:
Meh, this feels really hackish. Let's please not rummage around in a foreign object's internal state. The aim of this commit is to give Router *more* control about endpoints, and this does the opposite. Can this get a `router.add_endpoint()` or so?

Reply:
The original intention of the `Endpoint` object (and its location inside of `router.py`) was that it is `Endpoint` and `Router` are a single concern.  I nearly removed the `.drop_channel()` method on `Router` in favour of `Endpoint` reaching in to do it directly.  The only thing that stopped me is that we had trouble on closing already-closed channels in the past and I wanted to keep the extra debugging output in that function...

Reply:
Found the commit where I introduced this concept: https://github.com/cockpit-project/cockpit/commit/ba29116a7ce6ad9ea593bb919c54330736cbd72a

Reply:
OK, I leave this up to you. Resolving these two threads, that just leaves the better justification/comment on the remaining thread above.